Add Physio object generation from BIDS

maestroque opened the pull request:
<!-- Write all of the issues that are linked to this pull request. -->
<!-- If this PR is enough to close them you can write something like "Closes #314 and closes #42" -->
<!-- If you just want to reference them without closing them, you can add something like "References #112" -->
Closes #3 

<!-- Add a short description of the PR content here-->


## Proposed Changes
<!-- List major points of changes here, so that the reviewers can have a bit more context while looking at your work! -->
  - Add a function to load BIDS structures and automatically generate Physio objects
  - Add an MRI data object to store trigger parameters, universal to all generated Physio objects

## Change Type
<!-- Indicate the type of change you think your pull request is -->
- [ ] `bugfix` (+0.0.1)
- [X] `minor` (+0.1.0)
- [ ] `major`  (+1.0.0)
- [ ] `refactoring` (no version update)
- [ ] `test` (no version update)
- [ ] `infrastructure` (no version update)
- [ ] `documentation` (no version update)
- [ ] `other`

## Checklist before review
<!-- If this section is not clear, please read this part of the docs: https://phys2bids.readthedocs.io/en/latest/contributorfile.html#pr -->
<!-- You're invited to open a draft PR ASAP, but before marking it "ready for review", check that you have done the following: -->
- [ ] I added everything I wanted to add to this PR.
- [ ] \[Code or tests only\] I wrote/updated the necessary docstrings.
- [ ] \[Code or tests only\] I ran and passed tests locally.
- [ ] \[Documentation only\] I built the docs locally.
- [x] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
- [x] My code respects the adopted style, especially linting conventions.
- [x] The title of this PR is explanatory on its own, enough to be understood as part of a changelog.
- [ ] I added or indicated the right labels.
<!-- If relevant, you can add a milestone label or indicate an ideal timeline for this PR, including whether the progress of this PR is linked to other PRs being completed before or after it. -->
- [ ] I added information regarding the timeline of completion for this PR.
<!-- If you want, you can ask for reviews on a draft PR -->
- [ ] Please, comment on my PR while it's a draft and give me feedback on the development!

maestroque committed:
Add preliminary load_from_bids function

m-miedema commented:
See my [comment](https://github.com/physiopy/physutils/issues/3#issuecomment-2222840472) on the corresponding issue, I think we'll need to add more flexibility to column naming for this to be useful.

maestroque committed:
Add MRIConfig class

maestroque committed:
Fixes for working load_from_bids function

maestroque committed:
Remove supported columns and add detection for known columns

maestroque committed:
load_from_bids: Dictionary keys named after column names

maestroque commented:
@m-miedema it's fixed for the time being. I have implemented string literals searching, to give the physio_type property that is used in some functions.

maestroque committed:
load_from_bids: return trigger as Physio

m-miedema commented:
This looks reasonable to me! Let me know when it's ready for review :)

me-pic commented:
Hi @maestroque  ! For the addition of unit tests for those new functions, I've talked with a colleague and she recommended to just generate the structure on the fly when we run pytest (and generate random number for the timeseries if we even need to), if we are just concerned about the data structure ! If we are more interested about the content of those files and need actual real data you can take a look at that [BIDS app](https://giga-connectome.readthedocs.io/en/stable/contributing.html#running-the-demo)

maestroque commented:
Thanks @me-pic ! Sounds reasonable to me

me-pic commented on physutils/io.py:
```diff
@@ -8,13 +8,124 @@
+
+    for col in columns:
+        col_physio_type = None
+        if any([x in col for x in ["cardiac", "ppg", "ecg", "card"]]):
```
@maestroque Would it be possible to change for `col.lower()` ?

maestroque replied:
Nice catch!

me-pic reviewed (changes requested):
@maestroque Nice work on that PR !

me-pic commented on physutils/io.py:
```diff
@@ -8,13 +8,124 @@
 EXPECTED = ["data", "fs", "history", "metadata"]
 
 
+def load_from_bids(
```
Would it be possible to add the parameter `recording=None` ? The reason is that according to the current version of the BIDS spec physio data with different sampling rate need to be saved under different file names using the `recording` entity to differentiate them apart.

me-pic replied:
And `recording` should be added when you call `layout.get`

maestroque replied:
Got it, done!

me-pic commented on physutils/io.py:
```diff
@@ -8,13 +8,124 @@
+    if not op.exists(bids_path):
+        raise FileNotFoundError(f"Provided path {bids_path} does not exist")
+
+    layout = BIDSLayout(bids_path)
```
Not sure if there is a reason not to do that but I think we should specify `layout = BIDSLayout(bids_path, validate=False)`. I tested the current function with a BIDS dataset containing multiple physio files saved under names using different label values with the `recording` entity (ex. <match>_recording-pulse_<match> and <match>_recording-breathing_<match> ), and if `validate=False` is not specified, it just gives me an empty list after `layout.get()`

me-pic replied:
This change + the addition of the recording parameter would allow people with BIDS dataset containing physio data with different sampling rate to still be able to use those functionalities

maestroque committed:
Add BIDS structure creation test utils function

maestroque committed:
Add load_from_bids unit test

maestroque commented:
@me-pic the unit test is ready. I'll address your other comments shortly and it should be good

maestroque committed:
Review fixes

maestroque commented on setup.cfg:
I guess we should also change the `matplotlib >=3.9` line to non-constraining?

me-pic replied:
Yes, good catch !

maestroque committed:
Fix computed_metrics attribute initialization

maestroque committed:
Fix new_physio_like computed_metrics copying bug

me-pic commented on physutils/io.py:
```diff
@@ -8,13 +8,126 @@
+
+    for col in columns:
+        col_physio_type = None
+        if any([x in col.lower() for x in ["cardiac", "ppg", "ecg", "card"]]):
```
@m-miedema Do you think it would be reasonable to add "pulse" as a valid column name for "cardiac" ? And also "o2" and "co2" for "respiratory" ?

m-miedema replied:
Although O2 and CO2 reflect respiration, I don't think they should be labelled as respiratory! This is mainly because the respiratory response function should not be used for these types of data (see e.g. the fMRI literature on end-tidal response functions), and users may become confused with putting this data in the same category as a belt measurement.

For pulse, I've never personally seen it used as a naming convention, but why not!

me-pic commented on physutils/tests/utils.py:
```diff
@@ -77,3 +80,82 @@ def filter_physio(data, cutoffs, method, *, order=3):
+    run_id = "01"
+
+    bids_dir = pjoin(
+        data_dir, "bids-dir", f"sub-{subject_id}", f"ses-{session_id}", "func"
```
I think we should make sure that the test covers the possibility to have the recording entity mentioned in the filename. Ex. `sub-006_ses-03_task-test_run-01_recording-cardiac_physio.tsv.gz` would be a valid filename according to the BIDS standard. And since we also included a `recording` parameter in the `load_from_bids` function

me-pic replied:
Sorry @maestroque I think my comment was not clear. The tests should cover the possibility to have filenames with the `recording` entity as well as filenames that does not include that entity.

me-pic commented on physutils/physio.py:
```diff
@@ -220,10 +220,12 @@ def new_physio_like(
     physio_type = ref_physio.physio_type if copy_physio_type else None
-    computed_metrics = list(ref_physio.computed_metrics) if copy_computed_metrics else []
+    computed_metrics = (
+        dict(ref_physio.computed_metrics) if copy_computed_metrics else []
```
Should it be `dict(ref_physio.computed_metrics) if copy_computed_metrics else {}` instead ?

me-pic commented on physutils/physio.py:
```diff
@@ -542,3 +544,55 @@ def neurokit2phys(
         return cls(data, fs=fs, metadata=metadata, **kwargs)
+
+
+class MRIConfig:
```
Maybe I missed it, but is there any test to cover that class ?

maestroque replied:
No, it is not used yet, because it would need detection of the trigger channel in order to initialize it. I can open an issue defining the next steps for it, and I'll implement it if there is time after the workflow

me-pic commented on physutils/tests/test_io.py:
```diff
@@ -46,6 +50,24 @@ def test_load_physio(caplog):
+        # The data saved are the ones after t_0 = 0s
+        assert phys_array[col].data.size == 70000
+        assert phys_array[col].fs == 10000.0
+        assert phys_array[col].history[0][0] == "physutils.io.load_from_bids"
```
Would it be better to delete the bids dataset at the end of the function or just to add the `bids-dir` directory in the `.gitignore` ?

me-pic reviewed (commented):
Thank you @maestroque for integrating previous comments ! I've added some more :sweat_smile:  Should be pretty straight forward to integrate, but let me know if you need help !

m-miedema commented on physutils/io.py:
```diff
@@ -8,13 +8,126 @@
+
+    config_file = bids_file[0].get_metadata()
+    fs = config_file["SamplingFrequency"]
+    t_start = config_file["StartTime"]  # noqa
```
Is there a reason to define the start time here? It seems like it could be integrated into the time column checks, although its not necessary to do that in this PR. Perhaps open an issue to do a QA check and/or integrate the info into the specification of idx_0 in the event of a missing time column.

maestroque replied:
I integrated it into the `idx_0` definition

m-miedema commented on physutils/physio.py:
```diff
@@ -542,3 +544,55 @@ def neurokit2phys(
+    def __init__(self, slice_timings=None, n_scans=None, tr=None):
+        if np.ndim(slice_timings) > 1:
+            raise ValueError("Slice timings must be a 1-dimensional array.")
+        if np.size(slice_timings) != n_scans:
```
Here, the number of scans and the slice timings are actually two different things! The number of scans is the number of MRI volumes (ie. your number of fMRI time points), but the slice timing array should reflect the order in which each slice of a single volume was acquired. This could be something like e.g. 64 slices (you can see some sample cases [here](https://www.fil.ion.ucl.ac.uk/spm/docs/tutorials/fmri/preprocessing/slice_timing/)).

maestroque replied:
Yeap, you are right I missed this

m-miedema reviewed (changes requested):
Hi @maestroque ! This looks very nice and well-organized so far, I left a few comments to add to what @me-pic has included in her review. The main thing I want to understand better is the relationship between the MRIcfg and Physio objects, and whether it's relevant to extract any information from the trigger column itself (rather than the sidecar file). I don't think there's much in the MRI class that can be generated from the trigger column, other than the TR and number of scans, so while you can confirm this is correct from the sidecar, I don't think it's necessary to do so. I'm also wondering if you intend to e.g point the Physio object to its corresponding MRIcfg object in some way while you're setting them up - would this be helpful in metric calls later on?

maestroque committed:
Minor improvements

maestroque committed:
matplotlib dependency constraint removal

maestroque committed:
Integrate StartTime BIDS physio parameter

maestroque commented:
@me-pic I integrated your comments, so it should be good from that side!

@m-miedema I also addressed your comments on StartTime and n_slices. Regarding the use of the MRIConfig class and Physio, what you say is true, that was the intended use of the new class as we discussed in the past. It is pending the extraction of data from the trigger channel and the integration into the Physio object. I will open an issue about this, but I will tackle it if time allows after the workflow

maestroque commented:
Other than that, it should be gtg

maestroque committed:
Minor fix

me-pic commented on physutils/tests/utils.py:
```diff
@@ -77,3 +80,85 @@ def filter_physio(data, cutoffs, method, *, order=3):
+    df.to_csv(tsv_file, sep="\t", index=False, header=False, float_format="%.8e")
+
+    # Compress tsv file into tsv.gz
+    tsv_gz_file = pjoin(
```
I'm not sure I understand why the file is first saved as a tsp, then loaded and saved again as a compressed file. I think you can just directly save `df` as a `tsv.gz` since that is the required format according to the[ BIDS spec](https://bids-specification.readthedocs.io/en/stable/modality-specific-files/physiological-and-other-continuous-recordings.html).

maestroque replied:
Yes, you are right, it was left as such from debugging

me-pic reviewed (changes requested):
Thank you @maestroque for integrating the changes ! I've added to more comments, but once those are addressed I think we could merge 🚀

maestroque committed:
[test] BIDS structure creation utility function fix

maestroque commented:
@me-pic I think we are now ready, let me know if you have anything else in mind

me-pic reviewed (approved):
LGTM :tada:

m-miedema commented:
Great work, thanks @maestroque !

smoia commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`0.2.1`](https://github.com/physiopy/physutils/releases/tag/0.2.1) :rocket:
